### ISSUE TITLE
TS2339 url on ImportMeta

### ISSUE THREAD

**Issue by Bidek56** [denoland/deno#6564]
Running the folloing code:
`console.log(  (import.meta as ImportMeta).url)`
or 
`console.log(  import.meta.url)`

using: 

`deno run -c .\tsconfig.json .\meta.ts `

with [tsconfig.json](https://github.com/Bidek56/deno-py-websocket/blob/master/client/tsconfig.json)

results in:
```
error: TS2339 [ERROR]: Property 'url' does not exist on type 'ImportMeta'.
console.log(  (import.meta as ImportMeta).url)

```

but in globals.ts **ImportMeta** has url. Any help appreaciated. Thx

```
  interface ImportMeta {
    url: string;
    main: boolean;
  }
```

**Comment by Bidek56** [comment]
@kitsonk Would you know how to fix this error? Thx

**Comment by bartlomieju** [comment]
`ImportMeta` is defined in [`lib.deno.shared_globals.d.ts`](https://github.com/denoland/deno/blob/master/cli/js/lib.deno.shared_globals.d.ts) - if you add `deno.shared_globals` to `lib` in your config it should work as expected.

**Comment by Bidek56** [comment]
Adding:

```
        "lib": [
            "deno.ns",
            "deno.shared_globals"
        ]
```
works
but when I add:
```
        "lib": [
            "dom",
            "dom.iterable",
            "deno.ns",
            "deno.shared_globals"
        ]
```

all sorts of conflict errors appear.
```
error: TS6200 [ERROR]: Definitions of the following identifiers conflict with those in another file: ImportExportKind, Module, Instance, Memory, TableKind, Table, ValueType, Global, CompileError, LinkError, RuntimeError, ReadableStreamReadResult, CountQueuingStrategy, ByteLengthQueuingStrategy, WritableStream, TransformStream, DOMException, BufferSource, BlobPart, Blob, File, FormDataEntryValue, FormData, HeadersInit, Headers, RequestInfo, RequestCache, RequestCredentials, RequestMode, RequestRedirect, ReferrerPolicy, BodyInit, RequestDestination, Request, ResponseType, Response, TextDecoder, TextEncoder, URLSearchParams, URL, MessageEvent, ErrorEvent, Worker, performance, Event, EventTarget, EventListenerOrEventListenerObject, CustomEvent, AbortController, AbortSignal
interface Account {
~~~~~~~~~
    at asset:///lib.dom.d.ts:25:1

    Conflicts are in this file.
    declare namespace WebAssembly {
    ~~~~~~~
        at asset:///lib.deno.shared_globals.d.ts:10:1

TS6200 [ERROR]: Definitions of the following identifiers conflict with those in another file: ImportExportKind, Module, Instance, Memory, TableKind, Table, ValueType, Global, CompileError, LinkError, RuntimeError, ReadableStreamReadResult, CountQueuingStrategy, ByteLengthQueuingStrategy, WritableStream, TransformStream, DOMException, BufferSource, BlobPart, Blob, File, FormDataEntryValue, FormData, HeadersInit, Headers, RequestInfo, RequestCache, RequestCredentials, RequestMode, RequestRedirect, ReferrerPolicy, BodyInit, RequestDestination, Request, ResponseType, Response, TextDecoder, TextEncoder, URLSearchParams, URL, MessageEvent, ErrorEvent, Worker, performance, Event, EventTarget, EventListenerOrEventListenerObject, CustomEvent, AbortController, AbortSignal
declare namespace WebAssembly {
~~~~~~~
    at asset:///lib.deno.shared_globals.d.ts:10:1

    Conflicts are in this file.
    interface Account {
    ~~~~~~~~~
        at asset:///lib.dom.d.ts:25:1

TS2717 [ERROR]: Subsequent property declarations must have the same type.  Property 'byobRequest' must be of type 'ReadableStreamBYOBRequest', but here has type 'undefined'.
  readonly byobRequest: undefined;
           ~~~~~~~~~~~
    at asset:///lib.deno.shared_globals.d.ts:347:12

    'byobRequest' was also declared here.
        readonly byobRequest: ReadableStreamBYOBRequest | undefined;
                 ~~~~~~~~~~~
        at asset:///lib.dom.d.ts:12477:14

TS2375 [ERROR]: Duplicate number index signature.
  [index: number]: string;
  ~~~~~~~~~~~~~~~~~~~~~~~~
    at asset:///lib.deno.shared_globals.d.ts:528:3
```

and I believe I need dom in order to get React working.

**Comment by kitsonk** [comment]
Yeah, shared_globals is likely going to be a mess... an easier thing to do is in you main module and don't import `deno.shared_globals` in your lib.

```ts
declare global {
  interface ImportMeta {
    url: URL;
  }
}
```

**Comment by Bidek56** [comment]
```
declare global {
  interface ImportMeta {
    url: string;
  }
}
```
works on my code sample but **import.meta.url** is used everywhere in the code base, so this can not be a global solution.

**Comment by kitsonk** [comment]
It is a global solution, a module that contains it is always part of the program, that is why I suggested your main module, or the module that loads the library, or whatever...  This is core functionality of TypeScript: https://www.typescriptlang.org/docs/handbook/declaration-merging.html#global-augmentation

**Comment by Bidek56** [comment]
Would you know why [globals.ts](https://github.com/denoland/deno/blob/master/cli/js/globals.ts) is not imported by default since it's part of Deno cli/js?

**Comment by kitsonk** [comment]
That code is the internals of Deno.  It does get "imported" in the sense that is how all the runtime APIs are exposed within the Deno runtime.  What gets exposed is described in [lib.deno.shared_globals](https://github.com/denoland/deno/blob/master/cli/js/lib.deno.shared_globals.d.ts).  `globals.ts` imports it so we can make sure we align the APIs properly to what our runtime contract is.  The problem is that, for various reasons, the `deno.shared_globals` conflicts with a lot of what is in the `dom` lib.  `dom` describes a lot of things that Deno doesn't support (but most browsers do).  So if you are writing code that runs under Deno and you include `dom` and `deno.ns` manually, you might write code that the compiler might not flag as being an error, but will break at runtime, though in order to use certain libraries which assume they are running in a browser, like React, then you need to have the `dom` library loaded to support things like SSR.

Basically that file is meaningless to an end user of Deno.

All that being said, we _might_ want to move the Deno `ImportMeta` extensions into `deno.ns` (plus any other Deno specific augmentations that would vary from what is provided in the default `dom` lib).

**Comment by Bidek56** [comment]
This make sense, thanks for your help!!
